File: CPP/Lesson_228_lfu_cache.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 228 -- LFU Cache
// Category   : Linked List
// Difficulty : Hard
// Study Plan : Day 114
// =============================================================
//
// QUESTION:
//   LFU eviction; tie-break by least recently used. Use freq buckets of OrderedDict.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class LFU{int c=0,mn=0;unordered_map<int,pair<int,int>> kv;unordered_map<int,list<int>> f;unordered_map<int,list<int>::iterator> it;public:LFU(int c):c(c){}void bump(int k){int fr=kv[k].second;f[fr].erase(it[k]);if(f[fr].empty()&&mn==fr)mn++;f[fr+1].push_back(k);it[k]=prev(f[fr+1].end());kv[k].second++;}int get(int k){if(!kv.count(k))return -1;bump(k);return kv[k].first;}void put(int k,int v){if(c==0)return;if(kv.count(k)){kv[k].first=v;bump(k);return;}if((int)kv.size()>=c){int ek=f[mn].front();f[mn].pop_front();kv.erase(ek);it.erase(ek);}kv[k]={v,1};f[1].push_back(k);it[k]=prev(f[1].end());mn=1;}};
int main(){LFU c(2);c.put(1,1);c.put(2,2);cout<<c.get(1)<<"\n";c.put(3,3);cout<<c.get(2)<<"\n";}

File: Python/Lesson_228_lfu_cache.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 228 -- LFU Cache
# Category   : Linked List
# Difficulty : Hard
# Study Plan : Day 114
# =============================================================
#
# QUESTION:
#   LFU eviction; tie-break by least recently used. Use freq buckets of OrderedDict.
# =============================================================
from collections import defaultdict, OrderedDict
class LFU:
    def __init__(s,c): s.c=c; s.k={}; s.f=defaultdict(OrderedDict); s.mn=0
    def _bump(s,k):
        f=s.k[k][1]; v=s.f[f].pop(k)
        if not s.f[f] and s.mn==f: s.mn+=1
        s.f[f+1][k]=v; s.k[k]=(v,f+1)
    def get(s,k):
        if k not in s.k: return -1
        s._bump(k); return s.k[k][0]
    def put(s,k,v):
        if s.c==0: return
        if k in s.k: s.k[k]=(v,s.k[k][1]); s.f[s.k[k][1]][k]=v; s._bump(k); return
        if len(s.k)>=s.c:
            ek,_=s.f[s.mn].popitem(last=False); del s.k[ek]
        s.k[k]=(v,1); s.f[1][k]=v; s.mn=1
if __name__=="__main__":
    c=LFU(2); c.put(1,1); c.put(2,2); print(c.get(1)); c.put(3,3); print(c.get(2))
